src/agent/src/netlink.rs

```diff
@@ -422,6 +466,162 @@ impl Handle {
         Ok(())
     }
 
+    async fn add_rules<I>(&mut self, list: I) -> Result<()>
+    where
+        I: IntoIterator<Item = Rule>,
+    {
+        // Import rtnetlink::packet::nlas::rule::Nal objects that make sense only for this function
+        use packet::nlas::rule::Nla;
+
+        for rule in list {
+            let mut request = self
+                .handle
+                .rule()
+                .add()
+                .action(packet::constants::RTN_UNICAST);
+
+            if rule.priority >= 0 {
+                request
+                    .message_mut()
+                    .nlas
+                    .push(Nla::Priority(rule.priority as u32));
+            }
+
+            // The default value of mark is -1
+            if rule.mark >= 0 {
+                request
+                    .message_mut()
+                    .nlas
+                    .push(Nla::FwMark(rule.mark as u32));
+            }
+
+            // The default value of mask is -1
+            if rule.mask >= 0 {
+                request
+                    .message_mut()
+                    .nlas
+                    .push(Nla::FwMask(rule.mask as u32));
+            }
+
+            // The default value of goto is -1
+            if rule.goto >= 0 {
+                request = request.action(packet::constants::FR_ACT_GOTO);
+                request.message_mut().nlas.push(Nla::Goto(rule.goto as u32));
+            }
+
+            // The default value of flow is -1
+            if rule.flow >= 0 {
+                request.message_mut().nlas.push(Nla::Flow(rule.flow as u32));
+            }
+
+            // If Nla::Table value is set, request.table(value) will be overrided.
+            if rule.table >= 256 {
+                request
+                    .message_mut()
+                    .nlas
+                    .push(Nla::Table(rule.table as u32));
+            }
+
+            // The default value of suppressPrefixlen and suppressIfgroup is -1
+            if rule.table > 0 {
+                if rule.suppressPrefixlen >= 0 {
+                    request
+                        .message_mut()
+                        .nlas
+                        .push(Nla::SuppressPrefixLen(rule.suppressPrefixlen as u32));
+                }
+                if rule.suppressIfgroup >= 0 {
+                    request
+                        .message_mut()
+                        .nlas
+                        .push(Nla::SuppressIfGroup(rule.suppressIfgroup as u32));
+                }
+            }
+
+            let table = if rule.table > 0 && rule.table < 256 {
+                rule.table as u8
+            } else {
+                packet::constants::RT_TABLE_UNSPEC
+            };
+
+            // table will be override by Nla::Table
+            request = request.table(table);
+            match rule.get_family() {
+                IPFamily::v6 => {
```

Comment:
I noticed that there is too much duplicated code here. Could you encapsulate it into a method?